<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
   <TITLE> [rabbitmq-discuss] AlreadyClosedException always sets hardError	and initiatedByApplication to true
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:rabbitmq-discuss%40lists.rabbitmq.com?Subject=Re%3A%20%5Brabbitmq-discuss%5D%20AlreadyClosedException%20always%20sets%20hardError%0A%09and%20initiatedByApplication%20to%20true&In-Reply-To=%3C2C78313F-BB0B-4882-8183-290131293E8A%40rabbitmq.com%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="031228.html">
   <LINK REL="Next"  HREF="031234.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[rabbitmq-discuss] AlreadyClosedException always sets hardError	and initiatedByApplication to true</H1>
    <B>Steve Powell</B> 
    <A HREF="mailto:rabbitmq-discuss%40lists.rabbitmq.com?Subject=Re%3A%20%5Brabbitmq-discuss%5D%20AlreadyClosedException%20always%20sets%20hardError%0A%09and%20initiatedByApplication%20to%20true&In-Reply-To=%3C2C78313F-BB0B-4882-8183-290131293E8A%40rabbitmq.com%3E"
       TITLE="[rabbitmq-discuss] AlreadyClosedException always sets hardError	and initiatedByApplication to true">steve at rabbitmq.com
       </A><BR>
    <I>Tue Oct 22 15:46:30 BST 2013</I>
    <P><UL>
        <LI>Previous message: <A HREF="031228.html">[rabbitmq-discuss] AlreadyClosedException always sets hardError	and initiatedByApplication to true
</A></li>
        <LI>Next message: <A HREF="031234.html">[rabbitmq-discuss] AlreadyClosedException always sets hardError and initiatedByApplication to true
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#31232">[ date ]</a>
              <a href="thread.html#31232">[ thread ]</a>
              <a href="subject.html#31232">[ subject ]</a>
              <a href="author.html#31232">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>On 22 Oct 2013, at 11:42, Steve Powell &lt;<A HREF="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">steve at rabbitmq.com</A>&gt; wrote:

&gt;<i> Issuing a 'real work' call after closing the channel, like channel.basicConsume(), throws an AlreadyClosedException, claiming that the connection is closed, but conn.isOpen() still returns true.
</I>
Well, it appears that the AlreadyClosedException produces the &quot;clean connection shutdown&quot; string when it is a hard error (which is correct). It is called by the (internal) 'channel.ensureIsOpen()' method (typically, if we are about to communicate across the channel).  But the reference in the exception is then to a Channel, and although the exception implies that the connection has been closed, it hasn't.  This appears to contradict the hard error flag: which is supposed to mean the connection is closed.

I guess this means that the bug is the connection isn't closed by the application initiated call to a closed channel.  

You could argue this particular channel error doesn't violate the protocol, so there is no reason to suppose that the connection would (or should) be closed, so it shouldn't be a hard error.  [The code comments imply that hardError==true implies that getReference() returns a Connection object, which is actually not true anyway.]

As this thread stated right at the beginning (see subject), there is no provision in the code for issuing an AlreadyClosedException without it being a hard error.  If we could do this, the meaning of hardError would be preserved.

As in the case of referring to a consumer that the client has no record of, the client code can often determine if the channel is already closed, and not risk any protocol violations before throwing the exception.  In these cases, it ought not to be a hard error: the connection can remain unaffected.

Steve Powell  [Cell: +44-7815-838-558] [RabbitMQ, Pivotal]
&#8220;L&#8217;enfer, c&#8217;est les autres.&#8221; Sartre
-------------- next part --------------
An HTML attachment was scrubbed...
URL: &lt;<A HREF="http://lists.rabbitmq.com/pipermail/rabbitmq-discuss/attachments/20131022/403d6f31/attachment.htm">http://lists.rabbitmq.com/pipermail/rabbitmq-discuss/attachments/20131022/403d6f31/attachment.htm</A>&gt;
</PRE>


































<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="031228.html">[rabbitmq-discuss] AlreadyClosedException always sets hardError	and initiatedByApplication to true
</A></li>
	<LI>Next message: <A HREF="031234.html">[rabbitmq-discuss] AlreadyClosedException always sets hardError and initiatedByApplication to true
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#31232">[ date ]</a>
              <a href="thread.html#31232">[ thread ]</a>
              <a href="subject.html#31232">[ subject ]</a>
              <a href="author.html#31232">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="https://lists.rabbitmq.com/cgi-bin/mailman/listinfo/rabbitmq-discuss">More information about the rabbitmq-discuss
mailing list</a><br>
</body></html>
